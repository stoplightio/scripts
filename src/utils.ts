import { ParsingToken } from '@oclif/parser/lib/parse';
import { execSync } from 'child_process';
import { sync as findUp } from 'find-up';
import * as fs from 'fs';
import * as path from 'path';

export const buildCommand = (
  baseCommand: string,
  {
    defaultArgs = {},
    rawArgs = [],
    flags = [],
  }: { defaultArgs?: any; rawArgs?: ParsingToken[]; flags?: string[] } = {},
) => {
  // TODO: use matcher function to match baseCommand.cmd for Windows?

  // pull binary off the front, append rest of args back later
  const [binary, ...commandArgs] = baseCommand.split(' ');

  let command = findUp(path.join('node_modules', '.bin', binary), {
    cwd: process.cwd(),
  });
  if (!command) throw new Error(`Unable to locate node_modules/.bin binary for ${binary}`);

  command = [command, ...commandArgs].join(' ');

  for (const arg of rawArgs) {
    if (!flags.includes(arg.input.substring(2))) command += ` ${arg.input}`;
  }

  for (const arg in defaultArgs) {
    if (!defaultArgs.hasOwnProperty(arg) || command.match(arg)) continue;
    command += ' ' + defaultArgs[arg];
  }

  return command;
};

export const runCommand = (command: string, { handleError = true } = {}) => {
  try {
    return execSync(command, {
      stdio: 'inherit',
      cwd: process.cwd(),
      env: { ...process.env },
    });
  } catch (error) {
    if (
      handleError &&
      typeof error === 'object' &&
      error !== null &&
      'status' in error &&
      (typeof error.status === 'number' || error.status === null)
    ) {
      return process.exit(error.status);
    } else {
      throw error;
    }
  }
};

export const getConfigFilePath = (name: string): string => {
  const filePath = path.resolve(process.cwd(), name);
  if (fs.existsSync(path.resolve(process.cwd(), name))) {
    return filePath;
  }

  return require.resolve(`@stoplight/scripts/${name}`);
};

export const buildPath = (...args: any) => {
  return path.resolve(process.cwd(), ...args);
};

export const sortObjKeys = (obj: object) => {
  return Object.keys(obj)
    .sort()
    .reduce((accumulator, currentValue) => {
      // @ts-expect-error
      accumulator[currentValue] = obj[currentValue];
      return accumulator;
    }, {});
};
