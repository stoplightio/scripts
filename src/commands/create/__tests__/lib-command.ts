import * as utils from '../../../utils';
import LibCommand, { IResponses } from '../lib';

describe('sl-scripts create:lib', () => {
  let shellCommands: string[] = [];
  let libInstance: LibCommand;

  beforeEach(() => {
    shellCommands = [];
    libInstance = new LibCommand([], {
      name: 'create:lib',
      version: '',
      channel: '',
      // @ts-ignore
      pjson: '',
    });
    jest.spyOn(utils, 'runCommand').mockImplementation(val => shellCommands.push(val));
    jest.spyOn(process, 'cwd').mockImplementation(() => '/mock');
  });

  afterEach(() => jest.restoreAllMocks());

  describe('createPackage', () => {
    it('should create the right object for a ts library', () => {
      const responses: IResponses = {
        name: 'stoplight',
        description: 'description...',
        author: 'support@stoplight',
        repository: 'https://github.com/stoplightio/scripts.git',
        react: false,
        ci: 'circle',
        license: 'apache2',
        quality: 'codeclimate',
      };

      expect(JSON.stringify(libInstance.createPackage(responses), null, 2)).toBe(
        JSON.stringify(
          {
            name: responses.name,
            version: '0.0.0',
            description: responses.description,
            keywords: [],
            sideEffects: false,
            homepage: 'https://github.com/stoplightio/scripts',
            bugs: 'https://github.com/stoplightio/scripts/issues',
            author: responses.author,
            repository: {
              type: 'git',
              url: responses.repository,
            },
            license: 'Apache-2.0',
            main: 'src/index.ts',
            files: ['**/*', '!__tests__'],
            engines: {
              node: '>=8.3.0',
            },
            scripts: {
              build: 'sl-scripts build',
              'build.docs': 'sl-scripts build:tsdoc',
              commit: 'git-cz',
              lint: 'sl-scripts lint',
              'lint.fix': 'yarn lint --fix',
              release: 'sl-scripts release',
              'release.docs': 'sl-scripts release:docs',
              'release.dryRun': 'sl-scripts release --dry-run --debug',
              test: 'sl-scripts test',
              'test.prod': 'yarn lint && yarn test --coverage --no-cache',
              'test.update': 'yarn test --updateSnapshot',
              'test.watch': 'yarn test --watch',
            },
            peerDependencies: {},
            dependencies: {},
            devDependencies: {
              '@stoplight/scripts': '1.x.x',
              typescript: '3.1.6',
            },
            'lint-staged': {
              '*.ts': ['yarn lint.fix', 'git add'],
            },
            husky: {
              hooks: {
                'commit-msg': 'commitlint -E HUSKY_GIT_PARAMS',
                'pre-commit': 'lint-staged',
              },
            },
            config: {
              commitizen: {
                path: 'node_modules/cz-conventional-changelog',
              },
            },
            commitlint: {
              extends: ['@commitlint/config-conventional'],
            },
            release: {
              extends: '@stoplight/scripts/release',
            },
          },
          null,
          2
        )
      );
    });

    it('should create the right object for a react library', () => {
      const responses: IResponses = {
        name: 'stoplight',
        description: 'description...',
        author: 'support@stoplight',
        repository: 'https://github.com/stoplightio/scripts.git',
        react: true,
        ci: 'circle',
        license: 'apache2',
        quality: 'codeclimate',
      };

      expect(JSON.stringify(libInstance.createPackage(responses), null, 2)).toBe(
        JSON.stringify(
          {
            name: responses.name,
            version: '0.0.0',
            description: responses.description,
            keywords: [],
            sideEffects: false,
            homepage: 'https://github.com/stoplightio/scripts',
            bugs: 'https://github.com/stoplightio/scripts/issues',
            author: responses.author,
            repository: {
              type: 'git',
              url: responses.repository,
            },
            license: 'Apache-2.0',
            main: 'src/index.ts',
            files: ['**/*', '!__tests__'],
            engines: {
              node: '>=8.3.0',
            },
            scripts: {
              build: 'sl-scripts build',
              'build.docs': 'build-storybook -c .storybook -o docs-auto',
              commit: 'git-cz',
              lint: 'sl-scripts lint',
              'lint.fix': 'yarn lint --fix',
              release: 'sl-scripts release',
              'release.docs': 'sl-scripts release:docs',
              'release.dryRun': 'sl-scripts release --dry-run --debug',
              storybook: 'start-storybook -p 9001',
              test: 'sl-scripts test',
              'test.prod': 'yarn lint && yarn test --coverage --no-cache',
              'test.update': 'yarn test --updateSnapshot',
              'test.watch': 'yarn test --watch',
            },
            peerDependencies: {
              react: '16.x.x',
            },
            dependencies: {
              '@stoplight/ui-kit': '1.x.x',
            },
            devDependencies: {
              '@stoplight/scripts': '1.x.x',
              '@types/react': '16.x.x',
              '@types/react-dom': '16.x.x',
              react: '16.x.x',
              'react-dom': '16.x.x',
              typescript: '3.1.6',
            },
            'lint-staged': {
              '*.ts': ['yarn lint.fix', 'git add'],
            },
            husky: {
              hooks: {
                'commit-msg': 'commitlint -E HUSKY_GIT_PARAMS',
                'pre-commit': 'lint-staged',
              },
            },
            config: {
              commitizen: {
                path: 'node_modules/cz-conventional-changelog',
              },
            },
            commitlint: {
              extends: ['@commitlint/config-conventional'],
            },
            release: {
              extends: '@stoplight/scripts/release',
            },
          },
          null,
          2
        )
      );
    });
  });
});
