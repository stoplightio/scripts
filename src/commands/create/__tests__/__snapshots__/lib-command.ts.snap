// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sl-scripts create:lib createPackage should create the right object for a react library: react-lib 1`] = `
"{
  \\"name\\": \\"stoplight\\",
  \\"version\\": \\"0.0.0\\",
  \\"description\\": \\"description...\\",
  \\"keywords\\": [],
  \\"sideEffects\\": false,
  \\"homepage\\": \\"https://github.com/stoplightio/scripts\\",
  \\"bugs\\": \\"https://github.com/stoplightio/scripts/issues\\",
  \\"author\\": \\"support@stoplight\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"https://github.com/stoplightio/scripts.git\\"
  },
  \\"license\\": \\"Apache-2.0\\",
  \\"main\\": \\"src/index.ts\\",
  \\"files\\": [
    \\"**/*\\",
    \\"!__tests__\\",
    \\"!__stories__\\"
  ],
  \\"engines\\": {
    \\"node\\": \\">=8.3.0\\"
  },
  \\"scripts\\": {
    \\"build\\": \\"sl-scripts build\\",
    \\"build.docs\\": \\"build-storybook -c .storybook -o docs-auto\\",
    \\"commit\\": \\"git-cz\\",
    \\"lint\\": \\"sl-scripts lint\\",
    \\"lint.fix\\": \\"yarn lint --fix\\",
    \\"release\\": \\"sl-scripts release\\",
    \\"release.docs\\": \\"sl-scripts release:docs\\",
    \\"release.dryRun\\": \\"sl-scripts release --dry-run --debug\\",
    \\"storybook\\": \\"start-storybook -p 9001\\",
    \\"test\\": \\"sl-scripts test\\",
    \\"test.prod\\": \\"yarn lint && yarn test --coverage --no-cache\\",
    \\"test.update\\": \\"yarn test --updateSnapshot\\",
    \\"test.watch\\": \\"yarn test --watch\\"
  },
  \\"peerDependencies\\": {
    \\"react\\": \\"16.x.x\\"
  },
  \\"dependencies\\": {
    \\"@stoplight/ui-kit\\": \\"1.x.x\\"
  },
  \\"devDependencies\\": {
    \\"@stoplight/scripts\\": \\"3.x.x\\",
    \\"@types/react\\": \\"16.x.x\\",
    \\"@types/react-dom\\": \\"16.x.x\\",
    \\"react\\": \\"16.x.x\\",
    \\"react-dom\\": \\"16.x.x\\",
    \\"typescript\\": \\"3.1.6\\"
  },
  \\"lint-staged\\": {
    \\"*.{ts,tsx}$\\": [
      \\"yarn lint.fix\\",
      \\"git add\\"
    ]
  },
  \\"husky\\": {
    \\"hooks\\": {
      \\"commit-msg\\": \\"commitlint -E HUSKY_GIT_PARAMS\\",
      \\"pre-commit\\": \\"lint-staged\\"
    }
  },
  \\"config\\": {
    \\"commitizen\\": {
      \\"path\\": \\"node_modules/cz-conventional-changelog\\"
    }
  },
  \\"commitlint\\": {
    \\"extends\\": [
      \\"@commitlint/config-conventional\\"
    ]
  },
  \\"release\\": {
    \\"extends\\": \\"@stoplight/scripts/release\\"
  }
}"
`;

exports[`sl-scripts create:lib createPackage should create the right object for a ts library: plain-lib 1`] = `
"{
  \\"name\\": \\"stoplight\\",
  \\"version\\": \\"0.0.0\\",
  \\"description\\": \\"description...\\",
  \\"keywords\\": [],
  \\"sideEffects\\": false,
  \\"homepage\\": \\"https://github.com/stoplightio/scripts\\",
  \\"bugs\\": \\"https://github.com/stoplightio/scripts/issues\\",
  \\"author\\": \\"support@stoplight\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"https://github.com/stoplightio/scripts.git\\"
  },
  \\"license\\": \\"Apache-2.0\\",
  \\"main\\": \\"src/index.ts\\",
  \\"files\\": [
    \\"**/*\\",
    \\"!__tests__\\",
    \\"!__stories__\\"
  ],
  \\"engines\\": {
    \\"node\\": \\">=8.3.0\\"
  },
  \\"scripts\\": {
    \\"build\\": \\"sl-scripts build\\",
    \\"build.docs\\": \\"sl-scripts build:tsdoc\\",
    \\"commit\\": \\"git-cz\\",
    \\"lint\\": \\"sl-scripts lint\\",
    \\"lint.fix\\": \\"yarn lint --fix\\",
    \\"release\\": \\"sl-scripts release\\",
    \\"release.docs\\": \\"sl-scripts release:docs\\",
    \\"release.dryRun\\": \\"sl-scripts release --dry-run --debug\\",
    \\"test\\": \\"sl-scripts test\\",
    \\"test.prod\\": \\"yarn lint && yarn test --coverage --no-cache\\",
    \\"test.update\\": \\"yarn test --updateSnapshot\\",
    \\"test.watch\\": \\"yarn test --watch\\"
  },
  \\"peerDependencies\\": {},
  \\"dependencies\\": {},
  \\"devDependencies\\": {
    \\"@stoplight/scripts\\": \\"3.x.x\\",
    \\"typescript\\": \\"3.1.6\\"
  },
  \\"lint-staged\\": {
    \\"*.{ts,tsx}$\\": [
      \\"yarn lint.fix\\",
      \\"git add\\"
    ]
  },
  \\"husky\\": {
    \\"hooks\\": {
      \\"commit-msg\\": \\"commitlint -E HUSKY_GIT_PARAMS\\",
      \\"pre-commit\\": \\"lint-staged\\"
    }
  },
  \\"config\\": {
    \\"commitizen\\": {
      \\"path\\": \\"node_modules/cz-conventional-changelog\\"
    }
  },
  \\"commitlint\\": {
    \\"extends\\": [
      \\"@commitlint/config-conventional\\"
    ]
  },
  \\"release\\": {
    \\"extends\\": \\"@stoplight/scripts/release\\"
  }
}"
`;
